File: scr/P125.R
is_palindrome <- function(num) {
    original <- num
    reversed_num <- 0
    while (num > 0) {
        reversed_num <- reversed_num * 10 + num %% 10
        num <- num %/% 10
    }
    return(original == reversed_num)
}

squared_sum <- function(N) {
    max_square <- floor(sqrt(N)) + 1
    sums <- rep(FALSE, N + 1)

    for (n in 1:max_square) {
        s <- 0
        for (m in n:max_square) {
            s <- s + m * m
            if (s > N) {
                break
            }
            if (m > n) {
                sums[s] <- TRUE
            }
        }
    }
    return(sums)
}

P125 <- function(N) {
    sums <- squared_sum(N)
    palindrome_sums <- rep(FALSE, N + 1)

    total <- 0
    for (i in 1:N) {
        if (sums[i] && is_palindrome(i) && !palindrome_sums[i]) {
            total <- total + i
            palindrome_sums[i] <- TRUE
        }
    }
    return(total)
}

cat(P125(100000000), "\n")

File: scr/P125.go
package main

import (
	"fmt"
	"math"
)

func isPalindrome(num int) bool {
	original := num
	reversedNum := 0
	for num > 0 {
		reversedNum = reversedNum*10 + num%10
		num /= 10
	}
	return original == reversedNum
}

func squaredSum(N int) []bool {
	maxSquare := int(math.Sqrt(float64(N))) + 1
	sums := make([]bool, N+1)

	for n := 1; n < maxSquare; n++ {
		s := 0
		for m := n; m < maxSquare; m++ {
			s += m * m
			if s > N {
				break
			}
			if m > n {
				sums[s] = true
			}
		}
	}
	return sums
}

func P125(N int) int64 {
	sums := squaredSum(N)
	palindromeSums := make([]bool, N+1)

	var total int64
	for i := 0; i <= N; i++ {
		if sums[i] && isPalindrome(i) && !palindromeSums[i] {
			total += int64(i)
			palindromeSums[i] = true
		}
	}
	return total
}

func main() {
	fmt.Println(P125(100000000))
}
